django/contrib/admin/views/autocomplete.py

```diff
@@ -0,0 +1,169 @@
+import sys
+
+from django.core.exceptions import PermissionDenied
+from django.core.paginator import EmptyPage, Page, PageNotAnInteger, Paginator
+from django.db import router
+from django.http import Http404, JsonResponse
+from django.utils.encoding import force_text
+from django.utils.functional import cached_property
+from django.views.generic.list import BaseListView
+
+
+class AutocompletePage(Page):
+    def has_next(self):
+        """
+        Return true until queryset is smaller than the requested page size.
+
+        This avoids calling calculating the whole page count.
+        """
+        return self.object_list.count() >= self.paginator.per_page
+
+
+class AutocompletePaginator(Paginator):
+    """
+    A Django Paginator for big data sets.
+
+    Because users mostly use the first pages, it doesn't matter
+    if the page count is correct or not, as long as it's high enough.
+
+    Count() queries can take time, because they want to be exact.
+
+    So this paginator just doesn't count.
+    """
+
+    def _check_object_list_is_ordered(self):
+        """
+        Remove ordering and the corresponding warning.
+
+        Ordering is very slow, and we are doing a search here anyways.
+        There is not big advantage ordering the result set, especially
+        considering the performance trade offs.
+        """
+        self.object_list.query.clear_ordering(force_empty=True)
+
+    def _get_page(self, *args, **kwargs):
+        return AutocompletePage(*args, **kwargs)
+
+    def validate_number(self, number):
+        """Validates the given 1-based page number but doesn't care about empty pages."""
+        try:
+            number = int(number)
+        except (TypeError, ValueError):
+            raise PageNotAnInteger('That page number is not an integer')
+        if number < 1:
+            raise EmptyPage('That page number is less than 1')
+        return number
+
+    @cached_property
+    def count(self):
+        """
+        Do NOT count the result set.
+
+        A count() or exit() filtered query can be very slow on large tables.
+        This means Select2 will do one more request that it needs to, but
+        that is a small trade off.
+        """
+        return sys.maxsize
+
+
+class AutocompleteJsonView(BaseListView):
+    """
+    View that handles requests from `.AutocompleteWidget`.
+
+    The view only supports HTTP's GET method.
+    """
+    paginator_class = AutocompletePaginator
+    paginate_by = 20
+    admin_site = None
+
+    @property
+    def model(self):
+        return self.field.remote_field.model
+
+    @property
+    def field(self):
+        return self.original_model._meta.get_field(self.field_name)
```

Comment:
Given that `model` also exists, it is weird that `field` refers to `original_model`. (Also where is `original_model` defined -- if nowhere then certainly tests seems to be missing). Maybe also rename to source & target model